controller/tests/ControllerTests.cpp

```diff
@@ -0,0 +1,545 @@
+      update_sink_(std::make_unique<TestUpdateSink>(controller_)),
+      stream_factory_(minifi::io::StreamFactory::getInstance(configuration_)) {
+    configuration_->set(minifi::Configure::controller_socket_host, "localhost");
+    configuration_->set(minifi::Configure::controller_socket_port, "9997");
+    configuration_->set(minifi::Configure::nifi_security_client_certificate, (minifi::utils::file::FileUtils::get_executable_dir() / "resources" / "minifi-cpp-flow.crt").string());
+    configuration_->set(minifi::Configure::nifi_security_client_private_key, (minifi::utils::file::FileUtils::get_executable_dir() / "resources" / "minifi-cpp-flow.key").string());
+    configuration_->set(minifi::Configure::nifi_security_client_pass_phrase, "abcdefgh");
+    configuration_->set(minifi::Configure::nifi_security_client_ca_certificate, (minifi::utils::file::FileUtils::get_executable_dir() / "resources" / "root-ca.pem").string());
+    configuration_->set(minifi::Configure::controller_ssl_context_service, "SSLContextService");
+    ssl_context_service_ = std::make_shared<controllers::SSLContextService>("SSLContextService", configuration_);
+    ssl_context_service_->onEnable();
+    controller_service_provider_ = std::make_unique<TestControllerServiceProvider>(ssl_context_service_);
+  }
+
+  void initalizeControllerSocket(const std::shared_ptr<c2::ControllerSocketReporter>& reporter = nullptr) {
+    if (connection_type_ == ConnectionType::SSL_FROM_CONFIGURATION) {
+      configuration_->set(minifi::Configure::nifi_remote_input_secure, "true");
+    }
+    controller_socket_protocol_ = std::make_unique<minifi::c2::ControllerSocketProtocol>(*controller_service_provider_, *update_sink_, configuration_, reporter);
+    if (connection_type_ == ConnectionType::SSL_FROM_SERVICE_PROVIDER) {
+      controller_service_provider_->setSsl();
+    }
+    controller_socket_protocol_->initialize();
+  }
+
+  std::unique_ptr<minifi::io::Socket> createSocket() {
+    if (connection_type_ == ConnectionType::UNSECURE) {
+      return stream_factory_->createSocket("localhost", 9997);
+    } else {
+      return stream_factory_->createSecureSocket("localhost", 9997, ssl_context_service_);
+    }
+  }
+
+  void setConnectionType(ConnectionType connection_type) {
+    connection_type_ = connection_type;
+  }
+
+ protected:
+  ConnectionType connection_type_ = ConnectionType::UNSECURE;
+  std::shared_ptr<minifi::Configure> configuration_;
+  std::shared_ptr<TestStateController> controller_;
+  std::unique_ptr<TestUpdateSink> update_sink_;
+  std::shared_ptr<minifi::io::StreamFactory> stream_factory_;
+  std::unique_ptr<minifi::c2::ControllerSocketProtocol> controller_socket_protocol_;
+  std::shared_ptr<controllers::SSLContextService> ssl_context_service_;
+  std::unique_ptr<TestControllerServiceProvider> controller_service_provider_;
+};
+
+#ifdef WIN32
+TEST_CASE("TestWindows", "[controllerTests]") {
+  std::cout << "Controller Tests are not supported on windows";
+}
```

Comment:
This could be just a code comment, and I would include the related helper classes in the `ifndef` block, to avoid unnecessary compilation.

Reply:
Moved OS check for Windows to CMakeLists.txt file in 03f64d39caf87072f2cb0063cb4ea164f0ffe29e